lldpd: Implement more setting options

marek22k opened the pull request:
The TanoWrt fork of OpenWrt also has an lldpd package, which has a - in my eyes - better configuration file with more options. This PR would include this improved configuration file.
One difference is that in TanoWrt the user is called `lldpd` instead of `lldp` and the option to set the management IP is called differently. I have changed this accordingly - so that it works with the previous package.

Additionally I added the LLDP MED Fast Start option as well as the use of the `-k` flag, which sends less software information. (https://forum.openwrt.org/t/possibility-to-stop-sending-the-revision-in-lldpd/)

If anyone has a better idea for naming the `lldpmed_fast_start_tx_interval` option, I am open.

Thanks to @systemcrash for the help with Git!

marek22k committed:
lldpd: Implement more setting options

With `configure system platform` you can configure the platform for the CDP protocol.
If you specify the flags for the legacy / non-LLDP protocols several times in succession, the protocol is also activated if no neighbor is found that speaks this protocol. With CDP this has the additional function to indicate the protocol version.
Furthermore, the option to use two flags to send less information has been added (`-k`, `-i`).
Further details can be found in the man pages.

Signed-off-by: Marek Küthe <m.k@mk16.de>

marek22k commented:
An alternative would be to use https://github.com/tano-systems/meta-tanowrt/blob/master/meta-tanowrt/recipes-networking/lldpd/lldpd/files/lldpd.init.

marek22k committed:
lldpd: Add improvements from Tano Systems

The OpenWrt fork from Tano Systems has created both a LuCi interface and an improved configuration file. With this it is possible to restart the service more efficiently and specify more settings options.
The original file can be found at https://github.com/tano-systems/meta-tanowrt/blob/master/meta-tanowrt/recipes-networking/lldpd/lldpd/files/lldpd.init.
I only changed the username lldpd to lldp and renamed ip_sys_mgnt_ip back to the previous configuration format for backwards compatibility.

Signed-off-by: Marek Küthe <m.k@mk16.de>

marek22k committed:
lldpd: Implementation of LLDP MED fast start and option to send less software information

There is the possibility of "LLDP Med Fast Start" (see Man Pages for a detailed description), which can be enabled in the lldpd configuration file or explicitly enabled with Interval.
Furthermore, you can use the flag "-k" to prevent excessive software information from being sent.
In this commit both have been implemented.

Signed-off-by: Marek Küthe <m.k@mk16.de>

marek22k committed:
lldpd: Remove unnecessary items from Tano Systems

The configuration file from Tano Systems, in contrast to the old configuration file from OpenWrt, created unused JSON functions both did not delete the lldpd socket on exit.

Signed-off-by: Marek Küthe <m.k@mk16.de>

marek22k committed:
lldpd: Improve the location parameter

Add the location parameter. This was previously only partially implemented, which meant that setting the parameter had no effect.
The location parameter can contain coordinates or addresses, for example.

Signed-off-by: Marek Küthe <m.k@mk16.de>

howels commented:
This would be great - I noticed that TanoWRT also included an LLDP Luci app which would be a great way of viewing neighbor information. 
https://github.com/tano-systems/luci-app-tn-lldpd

hauke commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -90,10 +102,12 @@ start_service() {
+	[ $enable_cdp -gt 0 ] && procd_append_param command "-$(printf 'c%.0s' $(seq 1 $enable_cdp))"
+	[ $enable_fdp -gt 0 ] && procd_append_param command "-$(printf 'f%.0s' $(seq 1 $enable_fdp))"
+	[ $enable_sonmp -gt 0 ] && procd_append_param command "-$(printf 's%.0s' $(seq 1 $enable_sonmp))"
+	[ $enable_edp -gt 0 ] && procd_append_param command "-$(printf 'e%.0s' $(seq 1 $enable_edp))"
```
These 4 options are all boolean, they are only 0 or 1 `"-$(printf 'e%.0s' $(seq 1 $enable_edp))"` does not change it.

It would also be better if you do this change in a separate commit.

marek22k replied:
Mhh, I can't find the code section. It seems that a better solution has already been found. (Especially with CDP, as there are quite a few possibilities there).

hauke commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -1,15 +1,23 @@
 #!/bin/sh /etc/rc.common
+# Copyright (C) 2018 Anton Kikin <a.kikin@tano-systems.com>
```
Please split this big commit into  multiple commits to make it easier to review the changes.

marek22k replied:
Mhh, I don't know much about Git and I didn't write the code myself, so I don't know how far I can do something about it. Alternatively, maybe (part of) the git history could be taken from the TanoWrt repo.

howels replied:
How do you want the PR split? This is focussed on lldpd and rewrites the init script so that it respects settings in /etc/config/lldpd correctly, whereas the previous version was ignoring uci setting for lldpd interface bindings.
I can help restructure the PR if you like - I'm interested in seeing this included and know Git. 
@hauke any guidance on acceptable commit sizes? I see there are already 5 commits dealing with `lldpd.init` and the changes are limited to the correct reading/writing of settings from /etc/config/lldpd.

systemcrash replied:
Just break it up into random commit sizes until @hauke is satisfied :)

I guess the idea is to attempt to capture some of the original commit messages and any combined wisdom there. If taking the Tanto history, rebase from those commits and slap your changes on top.

howels replied:
I think the commits already reflect the decisions made here so there's not a lot to add. It's really a rewrite of a single file so I'm surprised that maintainers wouldn't want this squashed to a single commit when they merge it.
Changes can be summarised as - remove redundant LLDP example location and allow init script to correctly interact with UCI configuration for binding to interfaces and settng management interface when starting LLDPD.

systemcrash replied:
Yeah, then it's just to get some approval. 

@jow- @hauke @hnyman ?

howels commented:
Any feedback on this? This PR fixes a current bug where the interfaces listed in `/etc/config/lldpd` are ignored with the current init script.

MaZderMind commented:
@hauke can you elaborate on how you want the PR to be split? If @marek22k won't find the time I would happily rework / split the PR, however given that the changes proposed in this PR is basicly a backport from https://github.com/tano-systems I find the commit history actually okay.

marek22k commented:
@PolynomialDivision @stintel Any news?

howels commented:
At this point I am running this updates init script package and the luci app PR on 9 different devices. Working well, would appreciate it very much if maintainers can review.

PolynomialDivision commented:
For reference: https://github.com/openwrt/luci/pull/6456

PolynomialDivision commented:
Please increment PKG_RELEASE. :)

marek22k commented:
> Please increment PKG_RELEASE.

Done

systemcrash commented:
🥳

howels commented:
Please can someone review this?

systemcrash commented:
@marek22k can you fix the conflicts? Might just need a re-base.

marek22k commented:
@systemcrash It seems that a new LLDP parameter has been added. I have now entered this accordingly in my fork.

Furthermore I used the chance and added the LLDP Location parameter again. (Also in the LuCi App)ain.

marek22k commented:
If someone has an idea how to add the "configure system capabilities enabled capabilities" option with a multiple selection in LuCi, I would be very open for it!

systemcrash commented:
Please refresh your commits so that they contain a commit message (for the last two commits you added) so the tests here pass.

marek22k commented:
> @marek22k Please refresh your commits so that they contain a commit message (for the last two commits you added) so the tests here pass.

I have now adjusted the commit messages.

systemcrash commented:
"this can contain coordination" ... you mean coordinates? x,y? Lat+Long?

systemcrash commented:
OK - please update commit message to say "coordinates" instead 😄

systemcrash commented:
OK, it seems we're battling the CI here: 

`Pull request should not include merge commits`

I think this means that you should not merge into your PR, but should rebase it instead.

marek22k commented:
I have limited knowledge of Git. Is there a tutorial or something like that how I can remove the merge commit, rebase the whole thing without losing changes?

systemcrash commented:
I think one of the easier ways is to drop the commit with the merge and add
your changes to a new commit on top. (Force push and check here, you should
see that this PR has conflicts).

Rename your branch to something else.

Check out main/master, rebase it to openwrt/master, make a new branch, and
cherry pick your saved commits onto the fresh master.

Google away for cherry picks or rebase as a starting point.

Once you've got a result on an identically named branch as this PR is on,
force push.

marek22k commented:
```
$git rebase upstream/main
Auto-merging package/network/services/lldpd/files/lldpd.init
CONFLICT (content): Merge conflict in package/network/services/lldpd/files/lldpd.init
error: could not apply 89fa427837... lldpd: Implement more setting options
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 89fa427837... lldpd: Implement more setting options
```
I would prefer if we could continue with the merge. I have a number of conflicts that I can't resolve. A merge commit would also better reflect the temporal progression. The only other thing I could offer is to do everything in one commit.

marek22k commented:
Thanks to @systemcrash for the tutorial! I hope now the commit history looks better again.

marek22k commented:
Why does something like https://github.com/openwrt/openwrt/pull/13497 not require six Approves and this PR does? @openwrt-bot Maybe you can merge this PR too?

systemcrash commented:
I figure it's because this repo is the core system and its protocols so
needs a more thorough review to prevent problems for the broader user base
while the luci stuff only affects GUI.

Your last two commits could be squashed to one if a param is removed then
added.

marek22k commented:
One other thing, Git somehow doesn't accept the patches anymore. Does anyone have an idea why?
```
$ git apply lldpd.patch
lldpd.patch:372: trailing whitespace.
	
error: patch failed: package/network/services/lldpd/Makefile:9
error: package/network/services/lldpd/Makefile: patch does not apply
error: patch failed: package/network/services/lldpd/files/lldpd.init:53
error: package/network/services/lldpd/files/lldpd.init: patch does not apply
error: patch failed: package/network/services/lldpd/files/lldpd.init:19
error: package/network/services/lldpd/files/lldpd.init: patch does not apply
error: patch failed: package/network/services/lldpd/files/lldpd.init:39
error: package/network/services/lldpd/files/lldpd.init: patch does not apply
error: patch failed: package/network/services/lldpd/files/lldpd.init:16
error: package/network/services/lldpd/files/lldpd.init: patch does not apply
error: patch failed: package/network/services/lldpd/files/lldpd.init:88
error: package/network/services/lldpd/files/lldpd.init: patch does not apply
```

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,82 @@ find_release_info()
+	local lldp_tx_interval
+	local lldp_tx_hold
+	local lldp_mgmt_ip
+	local lldp_agetntype
```
You'll need to fix this :)

marek22k replied:
Thanks for the hint!

systemcrash commented:
> One other thing, Git somehow doesn't accept the patches anymore. Does anyone have an idea why?

Without further info as to what you're trying to do where, upon what commit, take it at face value:

`patch does not apply`... [any more/to this commit/etc]

marek22k commented:
> Without further info as to what you're trying to do where, upon what commit, take it at face value:

I tried to build the packages for lldpd like in https://github.com/openwrt/luci/pull/6456#issuecomment-1685052033.

systemcrash commented:
Are you (trying to) applying patches which are already in your commits?
Lldpd has no patches to apply at build time.

systemcrash commented:
OK - I'm satisfied.  @hauke @PolynomialDivision Anyone else?

PolynomialDivision commented:
The commit message description is not sticking to the normal 72 chars per line limit. And the last commit description is not that easy to understand and has two signed off lines?
https://github.com/openwrt/openwrt/pull/13018/commits/abea26f278a701e841853d9c10ffb1e9eabe2b6e

marek22k commented:
I didn't want to "hide" in the last commit message that I had previously removed the location attribute. I can gladly improve the commit message again. (The fact that it has two signed-off lines is also because I combined two commits into one).

Regarding the 72 character limit: Is there a tool that autimatically breaks the line (Git or Vim)? For example, my email client does it automatically. Furthermore, the CI seems to have somehow not noticed it.

marek22k commented:
I have now adapted the commit message from the last commit. I hope it is now easier to understand.

systemcrash commented:
Any other reviewers?

howels commented:
> Any other reviewers?

Please have a look, this is a big step for lldp support

howels commented:
@PolynomialDivision and @hauke please can you check this again? We need 6 reviewers here so would really appreciate the core team have a quick look.

PolynomialDivision commented:
I would love to finally get this merged. But I don't understand the complexity of the commit messages. For example, why don't you combine `lldpd: Add improvements from Tano Systems` and `lldpd: Remove unnecessary items from Tano Systems`? Just use the first commit and write that you also fixed the socket deletion while working on it. And please don't write `This commit originally consisted of two commits:` and split the commit message. Just mention that you fixed the location parameter instead of writing that you deleted it, because it was broken and than added it fixed again. Or just do one single commit message in which you describe everything that you did. And please break the lines. So for me the changes are completely valid and it is just about the commit messages. btw thanks for all your effort. :)

stintel commented:
I'm going to be blunt here because this PR spam has been going on long enough. If you can't be bothered to read our [contribution guidelines](https://openwrt.org/submitting-patches), why should we bother reviewing your contribution? If people would actually read and follow, those, and [the ones from the kernel](https://www.kernel.org/doc/html/latest/process/submitting-patches.html), on which ours are based, 

And no, @PolynomialDivision, a single commit with all the changes is not acceptable, as per those guidelines. This PR does so many things that I would argue *more* commits are needed, each with a proper description of why the change is needed. This way, each commit can be easily reviewed by itself.

marek22k commented:
> Just use the first commit and write that you also fixed the socket deletion while working on it. And please don't write `This commit originally consisted of two commits:` and split the commit message.

Thanks for the hint! I will remove this from the commit message. I'm sometimes unsure how to phrase things. If you have better descriptions for other commits, I'm open.



> And no, @PolynomialDivision, a single commit with all the changes is not acceptable, as per those guidelines. This PR does so many things that I would argue _more_ commits are needed, each with a proper description of why the change is needed. This way, each commit can be easily reviewed by itself.

Please see the context here. I have adopted the changes made by TanoWrt.

stintel commented:
> Please see the context here. I have adopted the changes made by TanoWrt.

That doesn't matter. If you're making changes, they should be explained. I don't know TanoWrt, I don't know which changes they made to lldpd, I don't know why they made them, ... As a maintainer, I'd like to know these things before accepting changes to the package I'm maintaining.

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -339,6 +322,6 @@ reload_service() {
 
 stop_service() {
-	rm -rf ${LLDPD_RUN} 2>/dev/null
+	rm -rf ${LLDPD_RUN} $LLDPSOCKET
```
Perhaps the `2>/dev/null` redirect should be retained here?

marek22k replied:
Wouldn't this mean that if an error occurs when the service is stopped, the user would not know about it?

marek22k commented:
> And please break the lines.

Is there a tool for this? For example, my e-mail client does this automatically.

howels commented:
> > And please break the lines.
> 
> Is there a tool for this? For example, my e-mail client does this automatically.

Using the cli git messages can be written like so:
```
git commit -m '<Summary>' -m '<Description>' --signoff
```
If you place line breaks in the description before closing the quotes they should be captured.
The `--signoff` should take care of the contributor guideline requirement to identify yourself. Just make sure you have `git config user.name` and `git config user.email` set.
Hope this helps.

howels commented:
Should I have a pass a cleaning this up with a new PR? I don't want to cause offence to @marek22k but I think I can see what the objections are from the OpenWRT maintainers and probably resolve these.

marek22k commented:
@howels If you like. I would be very happy if the LLDP improvements were included.

howels commented:
> @howels If you like. I would be very happy if the LLDP improvements were included.
> 
> If I'm not mistaken, @systemcrash also wanted to look at / clean up the PR. It might make sense not to have two people doing the same job?!

Not sure how we can do this without creating a new PR, any ideas?

howels commented:
> @howels If you like. I would be very happy if the LLDP improvements were included.
> 
> If I'm not mistaken, @systemcrash also wanted to look at / clean up the PR. It might make sense not to have two people doing the same job?!

I think this commit follows the contributor guidelines and incorporates the suggestions above. Apologies if I've missed anything:
https://github.com/howels/openwrt/commit/044096c0af2f667aaa36b663c95be48f4b67076e

marek22k commented:
@howels You are welcome to open a new PR from my side and close this one if necessary. The main thing (for me) is that the LLDP settings are improved. In the end, it is irrelevant (to me) who creates the commits and the PR.

In the end, I only created the PR because no one else wanted to do it (see https://forum.openwrt.org/t/possibility-to-stop-sending-the-revision-in-lldpd/164487).

howels commented:
> @howels You are welcome to open a new PR from my side and close this one if necessary. The main thing (for me) is that the LLDP settings are improved. In the end, it is irrelevant (to me) who creates the commits and the PR.
> 
> In the end, I only created the PR because no one else wanted to do it (see https://forum.openwrt.org/t/possibility-to-stop-sending-the-revision-in-lldpd/164487).

I have created https://github.com/openwrt/openwrt/pull/14193 taking care to describe the commit and format as per the contributor guidelines. Please can I kindly ask for pointers if this is insufficient, I have read the guidelines and this appears to match the guidelines and the type of information I see in the current master branch commit history.

systemcrash commented:
Please close this

howels commented:
Note that the commits from the other PR are now in master (ignore the closed status) so the Luci App can be worked on now.
